### PR TITLE
Paper: Supporting Greater Interactivity in the IPython Visualization Ecosystem

### PR DESCRIPTION
**Pull request by WarmCyan**
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_: Charles Lindsey @cdlindsey 

_Reviewers_: 
- kuntao zhao @ktzhao 
- Rithwik @ritwit
- Juan Cabanela @juancab

### CODE REVIEW

File: papers/nathan_martindale/myst.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        - ipycytoscape
+        - traitlets
+        - ipyreact
+        - ipyvuetify
+        - barlowtwins
+        - jupyter
+    - rule: link-resolves
+      severity: ignore
+      keys:
+        - https://www.energy.gov/doe-public-access-plan
```
Comment (rowanc1) [inline papers/nathan_martindale/myst.yml:65]: I think that this is correct, but it doesn't seem to be working! I can try and investigate in the next day or so.

Comment (fwkoch) [reply papers/nathan_martindale/myst.yml:65]: Yes, this is correct - the `link-resolves` check just isn't actually respecting these rules...

Keep this as-is, I'll fix up the check, then it should go green!

Comment (rowanc1) [reply papers/nathan_martindale/myst.yml:65]: Rerunning the checks made them go green! ✅ Thanks @fwkoch.

**Comment by ktzhao** [comment]
It is a pleasure to review your paper. I have thoroughly enjoyed reading your work and find it innovative.

The paper follows a logical flow, explaining the importance of interactive visualization, the limitations of IPyWidgets, the benefits of IPyVuetify, and finally introducing IPyOverlay as a solution. It explains IPyWidgets, IPyVuetify, and how IPyOverlay addresses their limitations. The example use case effectively illustrates the library's practical application.

However, I noticed that some sentences are a little complex and may cause confusion for some readers. Simplifying these sentences by breaking them down into smaller parts could enhance readability and ensure that your key points are communicated clearly to a wider audience.

For instance,  in the Introduction section, the sentence: 
"This ability to interact with and visualize data plays a crucial role in many data science/analysis workflows by allowing users to manipulate and explore, thus enhancing comprehension, improving user engagement, and often providing a faster route to an intuitive understanding of the data."

Could be revised to:
“Interacting with and visualizing data is crucial in many data science /analysis workflows. It allows users to manipulate and explore data, which enhances understanding, improves engagement, and often leads to a quicker intuitive grasp of the data.”

Similarly, in the IPyWidgets section:
"Although there are limitations to using IPyWidgets alone, this library brings a lot of value to the scientific Python community and the data science Jupyter workflow with its ability to marry the two language ecosystems, each of which include many visualization tools and capabilities."

Could be revised to:
"IPyWidgets has limitations, but it's still a valuable tool for the scientific Python community and data science Jupyter workflows. It allows us to combine two language ecosystems with many visualization tools and capabilities."

**Comment by ritwit** [comment]
The paper looks good to me. The writing is clear and concise. I can appreciate the utility of IPyOverlay.

I am not an expert in visualizations. But should there be more details on the implementation of the library? 

There is a minor rendering issue in the caption of `code 4`.

**Comment by cbcunc** [comment]
> However, I noticed that some sentences are a little complex and may cause confusion for some readers. Simplifying these sentences by breaking them down into smaller parts could enhance readability and ensure that your key points are communicated clearly to a wider audience.

@ktzhao @ritwit If you would please, add line by line suggestions per the GitHub PR review process: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request

File: papers/nathan_martindale/detailsondemand.png
Comment (JuanCab) [inline papers/nathan_martindale/detailsondemand.png:1]: If you truly want to reproduce "Inception" you should revise this figure to be consistent with the final draft of the paper... but I did think it a clever Figure to include!

File: papers/nathan_martindale/ipywidgets_state.png
Comment (JuanCab) [inline papers/nathan_martindale/ipywidgets_state.png:1]: "Python" should be capitalized in the caption "JS model changes mirrored in Python attributes".

File: papers/nathan_martindale/main.md [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+the available libraries that build on top of IPyWidgets, highlighting
+IPyVuetify, which addresses most of the limitations of base IPyWidgets. Finally,
+we discuss a library we have open sourced called IPyOverlay and an example
+use case that motivated its design. 
+
+### Details on Demand
+
+Details on demand is one piece of the "visual information-seeking mantra"
+[@shneidermanEyesHaveIt1996], a paper describing an overall design approach to
+building interactive tooling. The general flow of interaction proposed by this
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: This mention of details on demand should really refer to the AUTHOR and not "a paper".
```suggestion
Details on demand is one piece of the "visual information-seeking mantra" of Schneiderman (1996)
 describing an overall design approach to building interactive tooling. 
 The general flow of interaction proposed by this
```

File: papers/nathan_martindale/main.md [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+mantra is "overview first, zoom and filter, then details-on-demand"
+[@shneidermanEyesHaveIt1996]. Overview
+first ideally starts with broader aggregate visualizations: something to gain a sense
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Redundant `first`
```suggestion
ideally starts with broader aggregate visualizations: something to gain a sense
```

File: papers/nathan_martindale/main.md [hunk 4]
```diff
@@ -0,0 +1,756 @@
...
+of the general scope, complexity, or shape of the data. From there a user can
+identify their own areas of interest to hone in on with whatever zoom and filter
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Redundant since user unlikely to identify other's areas of interest:
```suggestion
identify areas of interest to hone in on with whatever zoom and filter
```

File: papers/nathan_martindale/main.md [hunk 5]
```diff
@@ -0,0 +1,756 @@
...
+tools are applicable for their data's modality. Finally, details on demand
+means that the user can see or interact with more in-depth and specific
+information for pieces of their data, often in the form of a pop-up window or
+similar. This selective detail capability provides a flexible user experience
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: I do not like the phrasing "pop-up window or similar" since it doesn't say "similar WHAT".  Maybe it would be better to say "in the form of something like a pop-up window or inset graph"?  I just find the current phrasing gets in the way to understanding the authors' meaning.

File: papers/nathan_martindale/main.md [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+A figure demonstrating details on demand in action when a user hovers over the figure
+reference for @fig:detailsondemand. 
+:::
+
+
+
+## IPyWidgets
+
+IPyWidgets is a library that enables implementing user interactivity inside of a
+Jupyter Notebook/Jupyter Lab. A key challenge this framework has to overcome is
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: I don't think you need to specify both Jupyter Notebook and Jupyter Lab since Jupyter notebooks (the documents) run in different implementations of Jupyter (including Jupyter in VS Code).  I would suggest just
```suggestion
Jupyter Notebook. A key challenge this framework has to overcome is
```

File: papers/nathan_martindale/main.md [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+
+
+### Common Framework
+
+One of the most important benefits of IPyWidgets is that it provides a low-level
+framework abstracting away many of the specifics for how communication between a
+kernel, server, and frontend takes place. This framework can be used to design
+new widgets by providing implementations for the models and view code, without
+needing to reinvent the various messaging system mechanisms that solve the state
+synchronization problem discussed above. All of this has allowed many libraries
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: This is a pet peeve of mine, I don't think it is *necessary* to fix, but I do not like the use of spatial terms like "above" for prose, I prefer time terms like "previously".  But I do believe this spatial language is commonly accepted now, I suppose I should not fight it.
```suggestion
synchronization problem discussed previously. All of this has allowed many libraries
```

File: papers/nathan_martindale/main.md [hunk 8]
```diff
@@ -0,0 +1,756 @@
...
+event-driven; the user interacts with widgets/visualizations, and these events
+run code that update the interface. This is most often
+implemented by registering "observers" or event-handlers, which tells IPyWidgets to
+run a specified function every time a particular event is raised. In practice,
+for the default components provided by the IPyWidgets library, this is mostly limited to
+value changes (e.g., a slider is moved, the text in a textbox is changed, or an
+item in a drop-down menu is selected).
+
+These types of events are sufficient for basic interfaces, but often more complex types of
+interactions or responsiveness can greatly enhance the capabilities of a tool.
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: You kind of wedge in "or responsiveness" in this sentence and it made it hard to follow.  Furthermore, you address this later in the text whereas this paragraph focuses exclusively on other interactions like mouse events, just keep it on that:
```suggestion
interactions can greatly enhance the capabilities of a tool.
```

File: papers/nathan_martindale/main.md [hunk 9]
```diff
@@ -0,0 +1,756 @@
...
+it challenging to prototype or put together a one-off custom component in a
+project. Several libraries we discuss in the next section have improved this
+process substantially, but traditionally a custom component library was
+developed by starting from a Jupyter-provided cookie-cutter template
+[@ipywidgets_cookiecutter] in a separate project.
+
+One of the difficulties that comes with a project created in this way is the
+need to include the development infrastructure of both a Python and a JavaScript
+project. JavaScript toolchains and dependency management are somewhat
+notoriously complicated. Prototyping the component can be a slow process as
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Can you provide a citation for "somewhat notoriously complicated"?  If it is "notorious" it suggests it is "well known" and thus should be citable.  If not, consider rewording this.  I think the existence of cookie-cutter templates definitely suggests this is a complex task.

File: papers/nathan_martindale/main.md [hunk 10]
```diff
@@ -0,0 +1,756 @@
...
+the HTML can be bound to JavaScript variables, which makes it possible to easily
+automate the JavaScript model $\rightarrow$ view updates. IPyVuetify brings two main
+advantages: a large library of flexible premade components and the ability to
+easily define new components on the fly using Vue to simplify the resulting
+JavaScript code.
+
+### Vuetify Component Library
+
+IPyVuetify is an IPyWidgets port of the Vuetify component library, a set of more
+than 70 premade material design UI widgets. Most of these components support a
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Can you define "material design" or cite the term?

File: papers/nathan_martindale/main.md [hunk 11]
```diff
@@ -0,0 +1,756 @@
...
+fairly extensive set of attributes and properties to change both appearance and
+functionality, and more importantly each generally has multiple types of events
+that can be observed. A simple example would be a textbox form input component.
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Run on sentence with two ideas:  the support of extensive attributes/properties and multiple types of events to be observed.  Split it up:
```suggestion
functionality.  More importantly, each iPyVuetify widget generally has multiple types of events
that can be observed. A simple example would be a textbox form input component.
```

File: papers/nathan_martindale/main.md [hunk 12]
```diff
@@ -0,0 +1,756 @@
...
+In IPyWidgets, the `Text` widget can mostly only observe a value change when the
+user changes the text inside. The IPyVuetify `TextField` can additionally raise
+events for key presses, mouse clicks, the element gaining and losing focus, as
+well as any HTML event. This level of control can be useful, for example, when a
+text value change triggers a heavy processing task. The interaction is more
+effective when the computation is delayed until the field loses focus or the
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Here I think the proper word is "responsive" instead of "effective":
```suggestion
responsive when the computation is delayed until the field loses focus or the
```

File: papers/nathan_martindale/main.md [hunk 13]
```diff
@@ -0,0 +1,756 @@
...
+well to reducing boilerplate by allowing bindings in the HTML to directly refer
+to Python attribute traitlets, removing both the need to specify the attribute
+in the JavaScript model as well as the function to update the view from the
+model. We discuss this further in the section on [implementing IPyOverlay in
+IPyWidgets](#implementing-in-ipywidgets). @code:ipyvuetify exemplifies this,
+defining a custom button with a color change on toggle, where the colors can be
+modified from Python as well, shown in @fig:ipyvuetify_example. In this
+particular example, both the `@click` event and `:color` property directly bind
+to the Python attributes, and no explicit JavaScript is necessary. Finally,
+IPyVuetify components make it easy to directly call functions across the
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Can you tighten this a bit by stating 
```suggestion
IPyVuetify makes it easy to directly call functions across the
```

File: papers/nathan_martindale/main.md [hunk 14]
```diff
@@ -0,0 +1,756 @@
...
+event based on geological differences in the paths to those locations from the
+source. An ideal model would exhibit path invariance and represent these signals similarly. We use the
+constructed visualization tool along with metadata (information other than the
+timeseries the model was trained on, e.g. event location, depth, distance) to
+understand how well the training objective succeeds in learning invariance to
+the metadata attributes that are unrelated to the event source properties.
+Additionally, we use the dashboard to help end users understand
+how well the model generalizes to new sources and to explore uncertainty.
+
+The dashboard includes a UMAP plot of signals embedded by the model under
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Please define what I assume is an acronym, UMAP, the first time you use it.

File: papers/nathan_martindale/main.md [hunk 15]
```diff
@@ -0,0 +1,756 @@
...
+exploration. Each datapoint is a single observed seismic signal, and these
+data points are then clustered with DBSCAN. <!-- the key benefit from IPyOverlay
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Please define DBSCAN the first time you use it.  I know the term is common enough in machine learning but the submissions to the SciPy proceedings are meant to be cross-disciplinary.  You may also want to link to the original 1996 by Ester, Kriegel, Sander, and Xu on the algorithm.

File: papers/nathan_martindale/main.md [hunk 16]
```diff
@@ -0,0 +1,756 @@
...
+clustering and produce narrow distributions.
+
+
+### Implementing in IPyWidgets
+
+Within base IPyWidgets, there is no straightforward way of constructing a
+layout in which one widget is overlaid on top of another. Certain other libraries
+have limited functionality for this within their own widget instances, for
+example Matplotlib can display an axes within another axes, as can plotly, and
+many libraries support basic tooltips with arbitrary textual content. Others
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Again, several ideas in one run-on sentence that should be separated:
```suggestion
layout in which one widget is overlaid on top of another. Certain other libraries
have limited functionality for this within their own widget instances.  For
example Matplotlib and poorly can display an axes within another axes.  Furthermore,
many libraries support basic tooltips with arbitrary textual content. Others
```

File: papers/nathan_martindale/main.md [hunk 17]
```diff
@@ -0,0 +1,756 @@
...
+here -->
+These approaches are generally easier to design and work
+with, and abstract away or entirely avoid some of the messier CSS layout rules
+required to render HTML elements that do not follow an ordered top-to-bottom/left-to-right flow.
+Fundamentally, to enable this behavior, we implement
+custom IPyWidget containers that use specific CSS rules to control for
+nonlinear layouts, namely relying on absolute positioning (by setting
+`position: absolute`), and each child widget setting `left` and `top` CSS
+properties to have pixel-level control of its location relative to the parent
+container. A diagram demonstrating this design principle is shown in
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: I was confused by this sentence, specifically what is "this behavior"... Do you mean the ability to use the easier `HBox` and `VBox` grid layout or from the context of the sentence, the ability to have absolute positioning and (while you don't mention it) layers!  I would rework this to make it clear from the beginning what behavior you are referring to.

File: papers/nathan_martindale/main.md [hunk 18]
```diff
@@ -0,0 +1,756 @@
...
+declaratively constructing a bound list of child widgets each wrapped in custom
+HTML that reactively responds to list updates and other model property changes.
+A simplified example is shown in @code:container.
+
+
+
+```{code-block} html
+:label: code:container
+:linenos: true
+:caption: A simplified version of the Vue template for `OverlayContainer`. The `<div> tag with `v-for` reactively produces a `<div>` tag for each widget in the `children` list attribute within the Python model. Note, the style attribute is bound and references the position arrays on lines 21--22. Changing the position of each overlay widget is done by modifying the corresponding values in these arrays. We keep these arrays only on the JavaScript side to prevent undue traffic between the Python and JavaScript models from high-frequency events (e.g., mouse movements) that can cause significant lag in the interface.
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: This caption gets mangled when rendered from MyST, I think you forgot to close the quote around the first `<div>` and that mucked it up, but check.

File: papers/nathan_martindale/main.md [hunk 19]
```diff
@@ -0,0 +1,756 @@
...
+simply specifying pixel coordinates within the background, but a significant
+challenge is locations within dynamic content such as interactive plots that can
+be panned and scaled. The conceptual approach to solving this for many plotting
+libraries is to store the x/y data point the connection should point to, the
+initial ranges of any plot axes, and current pixel boundaries of the plot within
+the container. Then, listening for any events that indicate an axis range has
+changed, handler functions can recalculate pixel locations based on the new
+relative position of the x/y data point within the new range.
+
+Unfortunately, in practice the functions necessary to do this for different
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Missing a comma
```suggestion
Unfortunately, in practice, the functions necessary to do this for different
```

File: papers/nathan_martindale/main.md [hunk 20]
```diff
@@ -0,0 +1,756 @@
...
+plotting libraries can vary substantially as a result of implementation differences, and
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Again, two separate (albeit related) ideas in a sentence, so separate them:
```suggestion
plotting libraries can vary substantially as a result of implementation differences. Thus
```

File: papers/nathan_martindale/main.md [hunk 21]
```diff
@@ -0,0 +1,756 @@
...
+dynamic content inside of ipympl (Matplotlib) and plotly plots. The authors
+intend to continue expanding the set of supported visualization libraries for
+this particular feature.
+
+<!--These connections are rendered within an `OverlayContainer` as SVG lines
+(implementation details)-->
+
+### Context Menus
+
+A common feature in UI design across most platforms is the context menu.
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: Might be better to refer to "details on design" as a UI design paradigm/principle in this first sentence where it makes sense rather than wedging it the text later on in the paragraph.

File: papers/nathan_martindale/main.md [hunk 22]
```diff
@@ -0,0 +1,756 @@
...
+:label: fig:context_menu
+An activated context menu providing a selection of commands, which can be
+expanded well beyond the single possible action that can be taken from a left-click. 
+:::
+
+## Conclusion
+
+The prevalence and utility of Jupyter notebooks in data analysis and scientific
+workflows has prompted the development of a community of tools for interactive
+visualization. IPyWidgets, one of the main drivers and the backbone that
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md]: I think this sentence is wordy and gets the same idea across without "one of the main drivers and"
```suggestion
visualization. IPyWidgets, the backbone that
```

File: papers/nathan_martindale/references.bib
Comment (JuanCab) [inline papers/nathan_martindale/references.bib:1]: For some reason, when your references were rendered they same out a bit mangled, at least in my preview of the web page.  Notably:
1. authors were sometimes in LastName, FirstInitial format and sometimes FirstName LastName format.  Not sure if this can be easily fixed.
2. The reference for _matplotlib/ipympl: Release 0.9.4_ seems malformed with no space between the first authors First and Last names, one author referred to only as "Todd" and a reference to **npm-to-cdn-bot**.  You might want to check this reference and manually add it to your `references.bib`.  I am honestly not sure why this is listed in the references as I see no reference to it in the text or in your `references.bib`.

**Review by JuanCab — CHANGES_REQUESTED** [review]
This paper was genuinely interesting to read.  I had run into this issue of trying to have a matplotlib widget respond to show an inset of the region clicked, the authors went ahead and developed a solution! 

I like the fact that this paper doesn't try to operate in a vacuum by presenting IPyOverlay alone as a completed project. Instead, it provides a sufficient background into iPyWidgets, the challenges of details-on-demand GUI design presented by iPyWidgets and the development of both iPyVuetify and finally iPyOverlay as solutions to this GUI design goal.  This helps the reader really grasp not only what iPyOverlay is adding to the iPyWidgets ecosystem but also why those features are truly important for the development of better data exploration tools.

**Prose issues addressed in individual comments** 

The only consistent flaw in the prose in this paper is that the draft presented has a tendency toward wordy and run sentences.  In many cases, the run on sentence could easily have been broken up into distinct sentences, each presenting a distinct idea.  I tried to note as many of those wordy and run on sentences in the comments of my review.

**Prose issues not noted in individual comments** 

I did also run *proselint* on the existing draft and it caught 3 real issues, only one of which are likely worth addressing.  The correction worth addressing:

1. In line 148 of main.md: *proselint* suggests 'interact' instead of 'interact with each other' since the 'with each other' is implied and redundant.

The corrections that may not matter given the conversion of this document into HTML.

1. *proselint* noted inconsistent spacing after period (1 vs. 2 spaces).
2. *proselint* noted the use of straight quotes "" instead of quotes “”.

** Code review**

**Code Block 1**: No issues

**Code Block 2**: The use of ellipsis in the lines
```
parameter1 = ipw.FloatSlider(...)
parameter2 = ipw.IntText(...)
parameter3 = ipw.IntSlider(...)
```
and the mention of functions that were not actually in the provided in the event handler functions made the function untestable.  I  made changes to the code so that it can be run and would suggest something like
```
import ipywidgets as ipw

# initialize the widgets
graph1_out = ipw.Output()
graph2_out = ipw.Output()

parameter1 = ipw.FloatSlider(min=0, max=10, value=1)
parameter2 = ipw.IntText(min=0, max=10, value=1)
parameter3 = ipw.IntSlider(min=0, max=10, value=1)

# create the dashboard layout, combining all the widgets
dashboard = ipw.VBox([
    ipw.HBox([graph1_out, graph2_out]),
    ipw.HBox([parameter1, parameter2, parameter3]),
])
   
# event handler functions
def on_parameter1_change(change):
    graph1_out.clear_output(wait=True)
    with graph1_out:
        print(f"parameter 1 is {change['new']} and parameter 2 is {parameter2.value}.")
        
def on_parameter2_change(change):
    graph1_out.clear_output(wait=True)
    graph2_out.clear_output(wait=True)
    with graph1_out:
        print(f"parameter 1 is {parameter1.value} and parameter 2 is {change['new']}.")
    with graph2_out:
        print(f"parameter 2 is {change['new']} and parameter 3 is {parameter3.value}.")
        
def on_parameter3_change(change):
    graph2_out.clear_output(wait=True)
    with graph2_out:
        print(f"parameter 2 is {parameter2.value} and parameter 3 is {change['new']}.")

# attach event handlers
parameter1.observe(on_parameter1_change, names=["value"])
parameter2.observe(on_parameter2_change, names=["value"])
parameter3.observe(on_parameter3_change, names=["value"])

dashboard
```

Code Block 3: I would have included the code to display the button, which is shown in Figure 5 but not in the actual code.  Again, I made the changes myself and would suggest something like:
```
import traitlets
import ipyvuetify as v

class ColoredToggleButton(v.VuetifyTemplate):
    color1 = traitlets.Unicode("var(--md-light-blue-500)").tag(sync=True)
    color2 = traitlets.Unicode("#AA6688").tag(sync=True)
    state = traitlets.Bool(True).tag(sync=True)

    @traitlets.default("template")
    def _template(self):
        return """
        <template>
            <v-btn 
                width="300px"
                :color="state ? color1 : color2"
                @click="state = !state"
            >Press to change color!</v-btn>
        </template>
        """

button = ColoredToggleButton()

# Display button
button 
```
followed by a mention of that running 
```
button.color2 = "var(--md-green-600)"
```
will change the second color programmatically.

Code Block 4: I was disappointed that we were not provided with example iPyOverlay code both to show an example implementation of the Python package (although we were shown the Vue template for an overlay container in Code 4, I wasn't hable to test it with the code provided)

File: papers/nathan_martindale/main.md (hunk 18 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md]: I sure did, that fixed it, thanks! (@ritwit I assume this is what your comment was about as well?)

**Comment by WarmCyan** [comment]
@JuanCab thanks for mentioning proselint, I didn't know this existed but I'll be making frequent use of it moving forward!

I updated programs 2 and 3 with your suggested changes, definitely agreed it's a lot nicer to be able to directly copy the snippet and have a functioning example in a notebook.

It's also a good point that we don't ever include an example of using ipyoverlay itself! I had previously been worried about running into the word count limit, but I think I'm quite a bit under. I've been wanting to work up a good concise example notebook for pop-out subplots to include in the repository itself, once I do that I'll add a code snippet for it in the paper

**Review by ktzhao — APPROVED** [review]

File: papers/nathan_martindale/detailsondemand.png
Comment (WarmCyan) [reply papers/nathan_martindale/detailsondemand.png:1]: Will do, thanks!

File: papers/nathan_martindale/main.md (hunk 7 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md]: Yeah that's totally fair, especially for like a PDF print I suppose "above" kinda loses any meaning if "above" is now an entirely different page! Fixed.

File: papers/nathan_martindale/main.md (hunk 9 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md]: Alas, my citations would probably only consist of reddit programming memes :grinning: I do think this is a fairly commonly pointed out issue, but I don't really have a good formal reference for it, so point taken! I reworded it instead to briefly explain why it can be complicated:
```
JavaScript toolchains and dependency management can quickly become 
complicated with the number of packages, build system tools, and config files 
often necessary.
```
(let me know if you think this is sufficient)

File: papers/nathan_martindale/main.md (hunk 10 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md]: Provided a reference to google's material design website

File: papers/nathan_martindale/main.md [hunk 23]
```diff
@@ -0,0 +1,756 @@
...
+that keeps the initial interface uncluttered without sacrificing the ability to
+organically explore the data.
+
+
+A simple and widespread example of details on demand are tooltips. Descriptions
+of components or data point values in visualizations would often create an
+unusably noisy interface if they were always visible. Instead, a more concise
+view is maintained while providing the necessary information in a spatially
+relevant view when the user specifically requests it. Readers viewing the web
+version of these proceedings can explore a simple example of details on demand
```
Comment (cdlindsey) [inline papers/nathan_martindale/main.md:80]: Do we see this "web version" in the curvenote preview?  Is this going to be a separate file?

File: papers/nathan_martindale/main.md [hunk 24]
```diff
@@ -0,0 +1,756 @@
...
+        
+def on_parameter3_change(change):
+    with graph2_out:
+        graph2_render_function(parameter2.value, parameter3.value)
+
+# attach event handlers
+parameter1.observe(on_parameter1_change, names=["value"])
+parameter2.observe(on_parameter2_change, names=["value"])
+parameter3.observe(on_parameter3_change, names=["value"])
+```
```
Comment (cdlindsey) [inline papers/nathan_martindale/main.md:201]: This example is pretty generic.  Could you change the names to make it have more meaningful function.  This would be easier to understand.

File: papers/nathan_martindale/main.md [hunk 25]
```diff
@@ -0,0 +1,756 @@
...
+data/attributes/values for a widget are stored) and a JavaScript _view_ (the
+code that creates the visual UI elements the user sees and interacts with in
+their browser). IPyWidgets connects to this backbone model via WebSockets,
+through which model changes and events are sent as serialized JSON.
+
+
+:::{figure} ipywidgets_state.png
+:label: fig:ipywidgets_state
+State mirroring between models and view interaction for an example slider widget.
+:::
```
Comment (cdlindsey) [inline papers/nathan_martindale/main.md:222]: The blue color in the bottom right arrow seems off.  Maybe you could make the bottom box a different shade of yellow and change the arrow as well.

File: papers/nathan_martindale/main.md [hunk 26]
```diff
@@ -0,0 +1,756 @@
...
+specific components and bridges to JavaScript visualization libraries. Other
+libraries such as Panel, Solara, and Streamlit provide more holistic
+integrations between IPyWidgets and other visualization techniques, 
+which can support production-style dashboards for
+use outside of Jupyter alone. Although these libraries address the limited component
+set included with IPyWidgets, another class of Python packages exists in this ecosystem
+that completely changes and improves the workflow for defining new widgets. These
+workflow improvements eliminate the need for JavaScript toolchains, and in some cases, they also
+provide the capabilities from various JavaScript web frameworks. A few
+we briefly touch on here are AnyWidget, IPyReact, and IPyVuetify.
```
Comment (cdlindsey) [inline papers/nathan_martindale/main.md:356]: Gradio probably deserves a mention too.

File: papers/nathan_martindale/main.md [hunk 27]
```diff
@@ -0,0 +1,756 @@
...
+are various layout widgets like navigation drawers, dialog windows, and
+steppers.
+
+<!-- would potentially be nice to see a searchbox for table --> 
+:::{figure} ipw_ipv.png
+:label: fig:ipw_ipv
+A comparison of a few IPyWidget components on the upper left and their corresponding
+IPyVuetify components on the upper right. The table on the bottom is a
+DataTable component, which does not have a default IPyWidgets equivalent.
+:::
```
Comment (cdlindsey) [inline papers/nathan_martindale/main.md:419]: As in "program 2", this would be more instructive if the buttons meant something.

File: papers/nathan_martindale/main.md (hunk 23 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md:80]: Yeah, the curvenote preview is the web version, so hovering over the "Figure 1" link pulls up a tooltip/details-on-demand example. That figure itself is then just a (almost recursive :grin:) screenshot of what hovering over the figure link looks like so that you still get the idea if reading on a non-web medium, like a PDF version of the paper.

File: papers/nathan_martindale/main.md (hunk 25 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md:222]: As in the blue arrow color clashes with the shade of yellow we're using or as in that arrow shouldn't be blue?

Our intent was to show the two pathways of changes colored based on source, so:
1. Interacting with the HTML/JS causes changes to propagate from the JS view -> JS model -> python model (represented as yellow arrows the whole way to match the JS yellow) 
2. Modifying values from python propagates to the view in the other direction, so the pathway is python model -> JS model -> JS view (which we represented with blue arrows the whole way to match the python blue)

I'm assuming the oddness is having the python blue arrow between two yellow JS boxes, and the idea was to show that that communication is being caused by a change on the python side. Does that make more sense or is there maybe a better way to represent this?

**Review by JuanCab — APPROVED** [review]

**Review by JuanCab — APPROVED** [review]

**Comment by JuanCab** [comment]
I apologize for the confusion, while I can review individual commits, apparently when I approve an individual commit, it just marks the entire PR as approved, not what I meant!

File: papers/nathan_martindale/main.md [hunk 28]
```diff
@@ -742,6 +742,67 @@ An activated context menu providing a selection of commands, which can be
 expanded well beyond the single possible action that can be taken from a left-click. 
 :::
 
+### Example Usage
+
+A minimal code example using IPyOverlay's containers and connection lines is
```
Comment (JuanCab) [inline papers/nathan_martindale/main.md:744]: Thank you for adding the code example!

**Review by JuanCab — COMMENTED** [review]
I have gone through all the commits through August 1, 2024 17:45 UTC and they all look good.

File: papers/nathan_martindale/main.md (hunk 23 above)
Comment (cdlindsey) [reply papers/nathan_martindale/main.md:80]: that makes sense, cool

File: papers/nathan_martindale/main.md (hunk 25 above)
Comment (cdlindsey) [reply papers/nathan_martindale/main.md:222]: gotcha, yeah, I guess a third color might be too confusing

Comment (JuanCab) [reply papers/nathan_martindale/main.md:222]: I'll note that for accessibility purposes (e.g. - usability by the color blind), the color should not be the only thing distinguishing the arrows if you want them to be distinct.  I don't have the figure here, but I thought I would mention it.  Doesn't directly address the question at hand, but just something I forgot to mention when originally reviewing the figure.

File: papers/nathan_martindale/references.bib
Comment (WarmCyan) [reply papers/nathan_martindale/references.bib:1]: I addressed the ipympl citation as best I could, (and made it more obvious, in the text where I reference it I had only called it interactive matplotlib) I think zenodo just grabbed the full list of github contributors and not all of them have a full name listed anywhere, but I at least removed the bot and fixed one of the names that it hadn't gotten correct. 

I'm also not really sure what to do about 1, I'm assuming there's something wrong in the bibtex syntax somewhere in the auto-collected doi ref, but I'm not sure how to debug that

File: papers/nathan_martindale/main.md (hunk 24 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md:201]: Yeah that's a good point, I renamed things to try to make it more obvious that this is showing a pattern where you can have parameters that influence multiple graphs, or ipywidget outputs that are influenced by multiple inputs that can be shared or have crossover

File: papers/nathan_martindale/main.md (hunk 26 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md:356]: Gradio is a good suggestion, I looked into it and was kind of surprised to see it doesn't actually have any direct ipywidgets compatibility. It has its own set of communication mechanisms, and while it can render inside of a notebook, as far as I can tell it's still entirely distinct and can't use other ipywidgets

File: papers/nathan_martindale/main.md (hunk 27 above)
Comment (WarmCyan) [reply papers/nathan_martindale/main.md:419]: The figure was just supposed to show the visual distinction between the types of components available in both, so I just labeled them according to the library they came from. I updated the caption to highlight these are just widget samples

File: papers/nathan_martindale/references.bib
Comment (JuanCab) [reply papers/nathan_martindale/references.bib:1]: Maybe this is a Curvenote issue?

**Review by JuanCab — COMMENTED** [review]
I like the code changes, they do make things a little clearer and easier to follow.

**Review by JuanCab — COMMENTED** [review]
I like the code changes, they do make things a little clearer and easier to follow.
